### PR TITLE
Added doc-gen feature.

### PR DESCRIPTION
**Pull request by RIg410**

### CODE REVIEW

File: dove/src/docs/options.rs [hunk 1]
```diff
@@ -0,0 +1,128 @@
+// Copyright (c) The Diem Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+use serde::{Deserialize, Serialize};
+
+/// Docgen options.
+#[derive(Deserialize, Serialize, Debug, Clone, PartialEq, Eq)]
+pub struct DocgenOptions {
```
Comment (vitvakatu) [inline dove/src/docs/options.rs:7]: It is not clear how it's different from original one. Could you clarify (here, in conversation)?

File: dove/src/cmd/build.rs [hunk 2]
```diff
@@ -97,25 +100,30 @@ impl Cmd for Build {
...
+                return Err(anyhow!("could not compile:{}", ctx.project_name()));
             }
             Ok(compiled_units) => {
                 // Verify and store compilation results...
                 self.verify_and_store(&ctx, files, compiled_units, &exclude_modules)?;
-                Ok(())
             }
         }
+
+        if let Some(env) = env {
```
Comment (vitvakatu) [inline dove/src/cmd/build.rs:122]: Is there any reason we need **not** to create GlobalEnv?

Comment (RIg410) [reply dove/src/cmd/build.rs:122]: GlobalEnv is used only for generating documentation and is not compatible with executor.

File: dove/src/docs/options.rs (hunk 1 above)
Comment (RIg410) [reply dove/src/docs/options.rs:7]: The original structure has an output_directory field and a doc_path . These fields are calculated and should not be available to the user for modification.

File: dove/src/cmd/docgen.rs [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+            ctx.dialect.as_ref(),
+            Some(sender),
+            StaticResolver::new(dep_list),
+        )
+            .build(&source_ref, true);
+
+        if let Some(env) = env {
+            Self::gen(&ctx, env)
+        } else {
+            Err(anyhow!(""))
```
[the remark this answers is not in the document]
Comment (RIg410) [reply dove/src/cmd/docgen.rs]: Done.

File: dove/src/cmd/build.rs (hunk 2 above)
Comment (vitvakatu) [reply dove/src/cmd/build.rs:122]: Okay, I'm just concerned about increasing code complexity because of additional if branches everywhere.

Comment (RIg410) [reply dove/src/cmd/build.rs:122]: This is a temporary version. We are changing the test system. After that, these conditions will be deleted.

**Review by vitvakatu — APPROVED** [review]

File: dove/src/cmd/docgen.rs [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+            ctx.dialect.as_ref(),
+            Some(sender),
+            StaticResolver::new(dep_list),
+        )
+        .build(&source_ref, true);
+
+        if let Some(env) = env {
+            Self::gen(&ctx, env)
+        } else {
+            unreachable!()
```
Comment (fzzr-) [inline dove/src/cmd/docgen.rs:51]: Is it really unreachable?

**Review by fzzr- — APPROVED** [review]
👍🔥 LGTM

Note for near future: Don't forget rid out of `GlobalEnv` where possible and `GlobalEnv` not needed.

**Pull request merged**